File: phenotype-skills/bindings/typescript/src/registry.ts
/**
 * Skill registry implementation
 */

import { SkillId, SkillManifest } from './types';
import { Skill } from './skill';

/**
 * Main registry for managing skills
 */
export class SkillRegistry {
  private readonly skills: Map<string, Skill> = new Map();

  /**
   * Register a new skill
   * @throws Error if skill is already registered
   */
  register(skill: Skill): void {
    const key = skill.id.toString();
    if (this.skills.has(key)) {
      throw new Error(`Skill already registered: ${skill.id}`);
    }
    this.skills.set(key, skill);
  }

  /**
   * Unregister a skill
   * @throws Error if skill not found
   */
  unregister(skillId: SkillId): Skill {
    const key = skillId.toString();
    const skill = this.skills.get(key);
    if (!skill) {
      throw new Error(`Skill not found: ${skillId}`);
    }
    this.skills.delete(key);
    return skill;
  }

  /**
   * Get a skill by ID
   * @throws Error if skill not found
   */
  get(skillId: SkillId): Skill {
    const key = skillId.toString();
    const skill = this.skills.get(key);
    if (!skill) {
      throw new Error(`Skill not found: ${skillId}`);
    }
    return skill;
  }

  /**
   * Try to get a skill by ID
   */
  tryGet(skillId: SkillId): Skill | undefined {
    return this.skills.get(skillId.toString());
  }

  /**
   * List all registered skills
   */
  list(): Skill[] {
    return Array.from(this.skills.values());
  }

  /**
   * Update a skill's manifest
   * @throws Error if skill not found
   */
  update(skillId: SkillId, manifest: SkillManifest): Skill {
    const key = skillId.toString();
    const skill = this.skills.get(key);
    if (!skill) {
      throw new Error(`Skill not found: ${skillId}`);
    }
    skill.updateManifest(manifest);
    return skill;
  }

  /**
   * Check if a skill exists
   */
  exists(skillId: SkillId): boolean {
    return this.skills.has(skillId.toString());
  }

  /**
   * Get the count of registered skills
   */
  get count(): number {
    return this.skills.size;
  }
}

/**
 * Builder for creating configured SkillRegistry instances
 */
export class SkillRegistryBuilder {
  /**
   * Build the registry
   */
  build(): SkillRegistry {
    return new SkillRegistry();
  }
}

File: phenotype-skills/bindings/python/src/phenotype_skills/registry.py
"""Skill registry implementation."""

from typing import Dict, List, Optional
from threading import RLock

from .types import SkillId, SkillManifest
from .skill import Skill


class SkillRegistry:
    """Main registry for managing skills."""

    def __init__(self):
        self._skills: Dict[SkillId, Skill] = {}
        self._lock = RLock()

    def register(self, skill: Skill) -> None:
        """Register a new skill."""
        with self._lock:
            if skill.id in self._skills:
                raise ValueError(f"Skill already registered: {skill.id}")
            self._skills[skill.id] = skill

    def unregister(self, skill_id: SkillId) -> Skill:
        """Unregister a skill."""
        with self._lock:
            if skill_id not in self._skills:
                raise KeyError(f"Skill not found: {skill_id}")
            return self._skills.pop(skill_id)

    def get(self, skill_id: SkillId) -> Skill:
        """Get a skill by ID."""
        with self._lock:
            if skill_id not in self._skills:
                raise KeyError(f"Skill not found: {skill_id}")
            return self._skills[skill_id]

    def try_get(self, skill_id: SkillId) -> Optional[Skill]:
        """Try to get a skill by ID."""
        with self._lock:
            return self._skills.get(skill_id)

    def list(self) -> List[Skill]:
        """List all registered skills."""
        with self._lock:
            return list(self._skills.values())

    def update(self, skill_id: SkillId, manifest: SkillManifest) -> Skill:
        """Update a skill's manifest."""
        with self._lock:
            if skill_id not in self._skills:
                raise KeyError(f"Skill not found: {skill_id}")
            skill = self._skills[skill_id]
            skill.update_manifest(manifest)
            return skill

    def exists(self, skill_id: SkillId) -> bool:
        """Check if a skill exists."""
        with self._lock:
            return skill_id in self._skills

    @property
    def count(self) -> int:
        """Get the count of registered skills."""
        with self._lock:
            return len(self._skills)


class SkillRegistryBuilder:
    """Builder for creating configured SkillRegistry instances."""

    def build(self) -> SkillRegistry:
        """Build the registry."""
        return SkillRegistry()
